rasterio/merge.py

```diff
@@ -298,27 +300,33 @@ def nullcontext(obj):
         nodataval = nodata
         logger.debug("Set nodataval: %r", nodataval)
 
+    inrange = False
     if nodataval is not None:
         # Only fill if the nodataval is within dtype's range
-        inrange = False
         if np.issubdtype(dt, np.integer):
             info = np.iinfo(dt)
-            inrange = (info.min <= nodataval <= info.max)
-        elif np.issubdtype(dt, np.floating):
-            if math.isnan(nodataval):
-                inrange = True
-            else:
+            inrange = info.min <= nodataval <= info.max
+        else:
+            if cmath.isfinite(nodataval):
                 info = np.finfo(dt)
-                inrange = (info.min <= nodataval <= info.max)
+                inrange = info.min <= nodataval <= info.max
+                nodata_dt = np.min_scalar_type(nodataval)
+                inrange = inrange & np.can_cast(nodata_dt, dt)
+            else:
+                inrange = True
+
         if inrange:
             dest.fill(nodataval)
         else:
             warnings.warn(
-                "The nodata value, %s, is beyond the valid "
-                "range of the chosen data type, %s. Consider overriding it "
-                "using the --nodata option for better results." % (
-                    nodataval, dt))
+                f"Ignoring nodata value. The nodata value, {nodataval}, cannot safely be represented "
+                f"in the chosen data type, {dt}. Consider overriding it "
+                "using the --nodata option for better results. "
+                "Falling back to first source's nodata value."
+            )
+            nodataval = first_nodataval
```

Comment:
This prevents a segmentation fault when using numpy 2.1.

Reply:
@snowman2 this must be the cause of the failure for GDAL 3.2. It's correct, so I'll mark the test an xfail for the oldest GDAL version.